core/remotes/docker/fetcher.go

```diff
@@ -318,19 +347,108 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
 			// Could use buffer pool here but this case should be rare
 			n, err := io.Copy(io.Discard, io.LimitReader(resp.Body, offset))
 			if err != nil {
+				if r.config.Limiter != nil {
+					r.config.Limiter.Release(parallelism)
+				}
 				return nil, fmt.Errorf("failed to discard to offset: %w", err)
 			}
 			if n != offset {
+				if r.config.Limiter != nil {
+					r.config.Limiter.Release(parallelism)
+				}
 				return nil, errors.New("unable to discard to offset")
 			}
 
+			// content range ignored, we can't do concurrent fetches here.
+			if parallelism > 1 && r.config.Limiter != nil {
+				r.config.Limiter.Release(parallelism - 1)
+			}
+			parallelism = 1
 		}
 	}
 
 	body := resp.Body
 	encoding := strings.FieldsFunc(resp.Header.Get("Content-Encoding"), func(r rune) bool {
 		return r == ' ' || r == '\t' || r == ','
 	})
+
+	totalSize, _ := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	remaining := totalSize - offset
+	if parallelism > 1 && chunkSize > 0 && remaining > minimumSizeForParallelDL && req.body == nil {
```

Comment:
Just FYI, here is similar piece of software to help clean up these conditions a bit - https://github.com/aws/aws-sdk-go-v2/blob/main/feature/s3/manager/download.go#L245

Reply:
Do you mean for inspiration ?